src/App.js

```diff
@@ -0,0 +1,116 @@
+import React, { useState, useCallback, useRef } from 'react';
+import Controls from './components/controls';
+import produce from 'immer';
+import './App.css';
+
+const neighborLocations = [
+  [0, 1],
+  [0, -1],
+  [1, -1],
+  [-1, 1],
+  [1, 1],
+  [-1, -1],
+  [1, 0],
+  [-1, 0]
+];
+
+function App() {
+
+  const [grid, setGrid] = useState();
+  const [cols, setCols] = useState();
+  const [rows, setRows] = useState();
+  const [running, setRunning] = useState(false);
+  const [speed, setSpeed] = useState(1000)
+
+  const generateEmptyGrid = (numRows, numCols) => {
+    const grid = Array.from({ length: numRows }).map(() => Array.from({ length: numCols }).fill(0))
+    setRows(numRows);
+    setCols(numCols);
+    setGrid(grid);
+  }
+
+  const runningRef = useRef(running);
+  runningRef.current = running;
+
+  const runSimulation = useCallback(() => {
+    if (!runningRef.current) {
+      return;
+    }
+    setGrid(grid => {
+      return produce(grid, gridCopy => {
+        for (let i = 0; i < rows; i++) {
+          console.log('$$$$$', grid[i]);
+          for (let k = 0; k < cols; k++) {
+            let neighbors = 0;
+            neighborLocations.forEach(([x, y]) => {
+              const newI = i + x;
+              const newK = k + y;
+              if (newI >= 0 && newI < rows && newK >= 0 && newK < cols) {
+                neighbors += grid[newI][newK];
+              }
+            });
+
+            if (neighbors < 2 || neighbors > 3) {
+              gridCopy[i][k] = 0;
+            } else if (grid[i][k] === 0 && neighbors === 3) {
+              gridCopy[i][k] = 1;
+            }
+          }
+        }
```

Comment:
This is crazy that you fit all of your logic and grid building all in this file. It does make it hard to follow though. Maybe this time constraints this was needed, but it would be nice if this more broken out and defined. Maybe even in separate files and sub-functions. Super clever though.